source/common/tracing/zipkin/span_context.cc

```diff
@@ -0,0 +1,131 @@
+#include "common/tracing/zipkin/span_context.h"
+
+#include "common/common/utility.h"
+#include "common/tracing/zipkin/zipkin_core_constants.h"
+
+namespace Zipkin {
+
+// The static functions below are needed due to C++ inability to safely concatenate static strings
+// belonging to different compilation units (the initialization order is not guaranteed).
+
+const std::string& SpanContext::FIELD_SEPARATOR() {
+  static const std::string* field_separator = new std::string(";");
+
+  return *field_separator;
+}
+
+const std::string& SpanContext::UNITIALIZED_SPAN_CONTEXT() {
+  static const std::string* unitialized_span_context =
+      new std::string("0000000000000000" + FIELD_SEPARATOR() + "0000000000000000" +
+                      FIELD_SEPARATOR() + "0000000000000000");
+
+  return *unitialized_span_context;
+}
+
+const std::string& SpanContext::HEX_DIGIT_GROUP_REGEX_STR() {
+  static const std::string* hex_digit_group_regex_str = new std::string("([0-9,a-z]{16})");
+
+  return *hex_digit_group_regex_str;
+}
+
+const std::string& SpanContext::SPAN_CONTEXT_REGEX_STR() {
+  // ^([0-9,a-z]{16});([0-9,a-z]{16});([0-9,a-z]{16})((;(cs|sr|cr|ss))*)$
+  static const std::string* span_context_regex_str = new std::string(
+      "^" + HEX_DIGIT_GROUP_REGEX_STR() + FIELD_SEPARATOR() + HEX_DIGIT_GROUP_REGEX_STR() +
+      FIELD_SEPARATOR() + HEX_DIGIT_GROUP_REGEX_STR() + "((" + FIELD_SEPARATOR() + "(" +
+      ZipkinCoreConstants::CLIENT_SEND + "|" + ZipkinCoreConstants::SERVER_RECV + "|" +
+      ZipkinCoreConstants::CLIENT_RECV + "|" + ZipkinCoreConstants::SERVER_SEND + "))*)$");
+
+  return *span_context_regex_str;
+}
+
+const std::regex& SpanContext::SPAN_CONTEXT_REGEX() {
+  static const std::regex span_context_regex(SPAN_CONTEXT_REGEX_STR());
+
+  return span_context_regex;
+}
+
+SpanContext::SpanContext(const Span& span) {
+  trace_id_ = span.traceId();
+  id_ = span.id();
+  parent_id_ = span.isSetParentId() ? span.parentId() : 0;
```

Comment:
These fields are not supposed to be `const`. The method populateFromString() changes the object state, i.e., it changes the values of those fields.